Results query parser

drothlis opened the pull request:
Extracted from #25. These are ready to merge now whereas #25 is going to take a while to get the results view up to par with the current static view.

Extends the http-service `/api/_private/results` endpoint with a `filtertext` parameter that takes a search string to limit the results returned. The string is in a format like this:

    testcase:example pass

which means: Match results where the testcase name contains "example" *and* where any field () contains "pass". You also use "or", parentheses, quotes for phrases with spaces in them, and dates with the "date" field. Currently only supports the fields we'd show in the results table: Date, Testcase, Result (pass/fail/error and failure reason string), and Duration. In future we can extend this to include the exception backtrace, and maybe even the logs.

**TODO**:

- [ ] Fix error message for duration:>5a.

drothlis commented:
On my PC, `test_that_a_completed_result_can_be_retrieved_via_results_interface` failed with:

```
    def test_that_a_completed_result_can_be_retrieved_via_results_interface(runner):
        job_id = run_tests(
            runner, '7292ddc0c46c2cf791a693889c7f2b720a585039',
            ['tests/example_tests.py::test_1']).json()['job_id']
        result = runner.get('/api/_private/results').json()['results'][0]
        assert (result['test_run']['test_case']['name']
                == "tests/example_tests.py::test_1")
>       assert result['job'] == {
            'category': 'default',
            'job_id': job_id,
            'job_uid': '/test-service-0123456789ab/0Xe4/%i' % job_id
        }
E       KeyError: u'job'
```

drothlis commented:
Fixed the test failure. This is ready to be reviewed / merged.

wmanley commented on _stbt_service/queryparser.py:
```diff
@@ -0,0 +1,164 @@
+            if fieldname == "result" and text == "error":
+                return {"test_run.exit_status": {"$gte": 2}}
+            # else
+            return {mongofieldname: re.compile(text, re.IGNORECASE)}
```
I guess we'll have to have a think about how we can implement `re.IGNORECASE` efficiently in the future,  but it shouldn't prevent efficient implementation.

wmanley commented:
I like how you've split out the commits copying the regex from Whoosh to the modifications.  A joy to review.

wmanley commented:
Review still incomplete, so possibly worth ignoring: It would be good to have some test-cases looking for things we explicitly don't support (because we don't think we can optimise)  e.g. "est" matching "test".

wmanley commented on http-service/service.py:
```diff
@@ -1368,10 +1420,14 @@ def json_array_response(iterator):
                               mimetype='application/json')
 
+    def json_response(dictionary):
+        """flask.jsonify doesn't handle pymongo `ObjectId`s."""
```
Perhaps we shouldn't bother including them, and use our own result id instead?

drothlis replied:
Probably a good idea. I'll leave dropping the ObjectId to the optimisation work. For now I'd even prefer to use `json_response` over `flask.jsonify` everywhere, for consistent encoding of dates. But all this will change when we move to bson so I'm going to leave as is for now, and we can change it in a future pull request.

wmanley commented:
Review complete.

drothlis commented:
> It would be good to have some test-cases looking for things we explicitly don't support (because we don't think we can optimise) e.g. "est" matching "test".

As this pull request stands "est" *does* match "test". I'd rather leave that change to the optimisation work (we'll just have to remember to do it before making a release).

drothlis commented:
* I rebased onto master
* Dropped 9c7bb25 (http-service /results: Limit the fields returned)
* Replaced the whitespace test (7ff3a80) with eebcd11 instead.
